File: leetcode/easy/valid_palindrome.py
"""
125. Valid Palindrome
https://leetcode.com/problems/valid-palindrome/

A phrase is a palindrome if, after converting all uppercase letters into lowercase letters 
and removing all non-alphanumeric characters, it reads the same forward and backward. 
Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.

Example 1:
Input: s = "A man, a plan, a canal: Panama"
Output: true
Explanation: "amanaplanacanalpanama" is a palindrome.

Example 2:
Input: s = "race a car"
Output: false
Explanation: "raceacar" is not a palindrome.

Example 3:
Input: s = " "
Output: true
Explanation: s is an empty string "" after removing non-alphanumeric characters.
Since an empty string reads the same forward and backward, it is a palindrome.

Constraints:
- 1 <= s.length <= 2 * 10^5
- s consists only of printable ASCII characters.
"""

import sys
import os

# Add the root directory to the Python path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))

from common.test_framework import run_tests


class Solution:
    def isAlphaNum(self,c):
            return_val = False
            if((ord('A') <= ord(c) <= ord('Z')) or
               (ord('a') <= ord(c) <= ord('z')) or
               (ord('0') <= ord(c) <= ord('9'))):
                return_val = True
            return return_val
    def isPalindrome(self, s: str) -> bool:
        
        # Solution 1
        """
        newStr = ""
        for c in s:
            if c.isalnum():
                newStr += c.lower()
        
        #reverse string
        reversedStr = newStr[::-1]
        if newStr == reversedStr:
            return True
        else:
            return False
        """
                
        # Solution 2
        

        
        L = 0
        R = len(s) - 1
        
        while(L < R):
            while((self.isAlphaNum(s[L]) == False) and (L < R)):
                L += 1
            while((self.isAlphaNum(s[R]) == False) and (R > L)):
                R -= 1
                
            if s[L].lower() != s[R].lower():
                return False
            L = L + 1
            R = R - 1
            
        return True

def test_solution():
    solution = Solution()
    
    test_cases = [
        (solution.isPalindrome, ["A man, a plan, a canal: Panama"], True, "Example 1"),
        (solution.isPalindrome, ["race a car"], False, "Example 2"),
        (solution.isPalindrome, [" "], True, "Example 3"),
        (solution.isPalindrome, ["12321"], True, "Numeric palindrome"),
        (solution.isPalindrome, ["0P"], False, "Mixed alphanumeric"),
        (solution.isPalindrome, ["ab_a"], True, "String with underscore"),
        (solution.isPalindrome, [""], True, "Empty string"),
        (solution.isPalindrome, ["!@#$"], True, "Special characters only"),
        (solution.isPalindrome, ["Race a Car"], False, "Case sensitivity test")
    ]
    
    run_tests(test_cases)

if __name__ == "__main__":
    test_solution()

File: java/leetcode/easy/ValidPalindrome.java
// 125. Valid Palindrome
// https://leetcode.com/problems/valid-palindrome/
//
// A phrase is a palindrome if, after lowercasing and removing non-alphanumeric
// characters, it reads the same forwards and backwards. (Ported from Python.)
package leetcode.easy;

import common.TestFramework;

public class ValidPalindrome {
    // Two-pointer approach. Time O(n), Space O(1).

    // Use Character.isLetterOrDigit(c) in the loop 


    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        if (s == " ")
        {
            return true;
        }
        while(left < right)
        {
            char currFirst = s.charAt(left);
            char currlast = s.charAt(right);
           while(!Character.isLetterOrDigit(currFirst) && (left < right))
            {
                left++;
                currFirst = s.charAt(left);
            } 
            while(!Character.isLetterOrDigit(currlast) && (left < right))
            {
                right--;
                currlast = s.charAt(right);
            } 
            
            if(Character.toLowerCase(currFirst) != Character.toLowerCase(currlast))
            {
                return false;
            }
            
            left++;
            right--;
        }
        
        return true;
    }

    public static void main(String[] args) {
        ValidPalindrome s = new ValidPalindrome();
        TestFramework t = new TestFramework();

        t.check("Example 1", s.isPalindrome("A man, a plan, a canal: Panama"), true);
        t.check("Example 2", s.isPalindrome("race a car"), false);
        t.check("Space only", s.isPalindrome(" "), true);
        t.check("Numeric palindrome", s.isPalindrome("12321"), true);
        t.check("Mixed alphanumeric", s.isPalindrome("0P"), false);
        t.check("Underscore skipped", s.isPalindrome("ab_a"), true);
        t.check("Empty string", s.isPalindrome(""), true);
        t.check("Special characters only", s.isPalindrome("!@#$"), true);

        System.exit(t.summary());
    }
}
